### DISCUSSION TITLE
usage/callbacks/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#17484] [Docs]
# usage/callbacks/

Explore Ultralytics callbacks for training, validation, exporting, and prediction. Learn how to use and customize them for your ML models.

https://docs.ultralytics.com/usage/callbacks/

**Comment by D3RK0** [comment]
Hi, in the documentation above, I see that `predictor.batch` is used, which I assume refers to the current batch of the predictor. Is there a similar `trainer.batch` for a trainer? I need it in order to apply data augmentation during the training of my YOLOv8 classifier.

**Reply by D3RK0** [reply]
I forgot to ask which (callback) event do I need to select?

**Reply by glenn-jocher** [reply]
@D3RK0 to apply data augmentation during training, you can use the `on_train_batch_start` callback to modify the data before each training batch is processed.

**Reply by D3RK0** [reply]
@glenn-jocher Many thanks for your quick response!

I am dealing with following problem though:

# Problem
When I do `trainer.batch`, I get the following error: 

```
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2022.3.3\plugins\python-ce\helpers\pydev\pydevd.py", line 1534, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2022.3.3\plugins\python-ce\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\Users\derk_\Documents\Universiteit\IRP\Research_Project\main.py", line 397, in <module>
    main()
  File "C:\Users\derk_\Documents\Universiteit\IRP\Research_Project\main.py", line 377, in main
    model.train(mode="classify",
  File "C:\Users\derk_\anaconda3\envs\cobra\Lib\site-packages\ultralytics\engine\model.py", line 802, in train
    self.trainer.train()
  File "C:\Users\derk_\anaconda3\envs\cobra\Lib\site-packages\ultralytics\engine\trainer.py", line 207, in train
    self._do_train(world_size)
  File "C:\Users\derk_\anaconda3\envs\cobra\Lib\site-packages\ultralytics\engine\trainer.py", line 368, in _do_train
    self.run_callbacks("on_train_batch_start")
  File "C:\Users\derk_\anaconda3\envs\cobra\Lib\site-packages\ultralytics\engine\trainer.py", line 168, in run_callbacks
    callback(self)
  File "C:\Users\derk_\Documents\Universiteit\IRP\Research_Project\main.py", line 185, in batch_augmentation_callback
    batch = trainer.batch  # Access the current batch
            ^^^^^^^^^^^^^
AttributeError: 'ClassificationTrainer' object has no attribute 'batch'
```

# My code
```
model = YOLO("yolov8m-cls.pt")

model.add_callback("on_train_batch_start", batch_augmentation_callback)

model.train(mode="classify",
            data=data_file_dir,
            epochs=args.n_epochs)

# Where batch_augmentation_callback is my callback function
def batch_augmentation_callback(trainer):
    batch = trainer.batch                                        # This code line causes the error reported above
    images, labels = batch['img'], batch['label']
    # More code below to create data augmentation
```

# Content of `trainer` object that is passed into `batch_augmentation_callback` (my callback function)

- trainer = {ClassificationTrainer} <ultralytics.models.yolo.classify.train.ClassificationTrainer object at 0x0000016929FA5010>
-  accumulate = {int} 4
-  amp = {bool} False
-  args = {IterableSimpleNamespace} IterableSimpleNamespace(task='classify', mode='train', model='yolov8m-cls.pt', data='C:/Users/derk_/Documents/Universiteit/IRP...ugment='randaugment', erasing=0.4, crop_fraction=1.0, cfg=None, tracker='botsort.yaml', save_dir='C:\\runs\\classify\\train53')
-  batch_size = {int} 16
-  best = {WindowsPath} WindowsPath('C:/runs/classify/train53/weights/best.pt')
-  best_fitness = {NoneType} None
-  callbacks = {defaultdict: 25} defaultdict(<class 'list'>, {'on_pretrain_routine_start': [<function on_pretrain_routine_start at 0x0000016929C79440>, <functi...C7A2A0>, <function on_export_start at 0x000001692A55C220>], 'on_export_end': [<function on_export_end at 0x0000016929C7A340>]})
-  csv = {WindowsPath} WindowsPath('C:/runs/classify/train53/results.csv')
-  data = {dict: 5} {'names': {0: '1', 1: '10', 2: '100', 3: '101', 4: '102', 5: '11', 6: '12', 7: '13', 8: '14', 9: '15', 10: '16', 11: '17', 12: '18', 13: '19', 14: '2', 15: '20', 16: '21', 17: '22', 18: '23', 19: '24', 20: '25', 21: '26', 22: '27', 23: '28', 24: '29', 25: 
-  device = {device} device(type='cpu')
-  ema = {ModelEMA} <ultralytics.utils.torch_utils.ModelEMA object at 0x000001692AC80C90>
-  epoch = {int} 0
-  epoch_time = {NoneType} None
-  epoch_time_start = {float} 1731355033.8569636
-  epochs = {int} 1
-  fitness = {NoneType} None
-  hub_session = {NoneType} None
-  last = {WindowsPath} WindowsPath('C:/runs/classify/train53/weights/last.pt')
-  loss = {NoneType} None
-  loss_names = {list: 1} ['loss']
-  metrics = {dict: 3} {'metrics/accuracy_top1': 0, 'metrics/accuracy_top5': 0, 'val/loss': 0}
-  model = {ClassificationModel} ClassificationModel(\n  (model): Sequential(\n    (0): Conv(\n      (conv): Conv2d(3, 48, kernel_size=(3, 3), stride=(2, 2), padd...1)\n      (drop): Dropout(p=0.0, inplace=True)\n      (linear): Linear(in_features=1280, out_features=102, bias=True)\n    )\n  )\n)
-  optimizer = {AdamW} AdamW (\nParameter Group 0\n    amsgrad: False\n    base_momentum: 0.85\n    betas: (0.9499811807836418, 0.999)\n    capturable: Fa...  lr: 0.004018066447703833\n    max_lr: 0.1\n    max_momentum: 0.95\n    maximize: False\n    min_lr: 4e-07\n    weight_decay: 0.0\n)
-  plot_idx = {list: 6} [0, 1, 2, -3465, -3464, -3463]
-  plots = {dict: 0} {}
-  resume = {bool} False
-  save_dir = {WindowsPath} WindowsPath('C:/runs/classify/train53')
-  save_period = {int} 1
-  scaler = {GradScaler} <torch.amp.grad_scaler.GradScaler object at 0x000001692921A0D0>
-  scheduler = {OneCycleLR} <torch.optim.lr_scheduler.OneCycleLR object at 0x000001692ACF5210>
-  start_epoch = {int} 0
-  stop = {bool} False
-  stopper = {EarlyStopping} <ultralytics.utils.torch_utils.EarlyStopping object at 0x000001692ACF7190>
-  stride = {int} 32
-  test_loader = {InfiniteDataLoader: 32} <ultralytics.data.build.InfiniteDataLoader object at 0x000001692AAD7B50>
-  testset = {WindowsPath} WindowsPath('C:/Users/derk_/Documents/Universiteit/IRP/Research_Project/Oxford-Flower-dataset/jpg/val')
-  tloss = {NoneType} None
-  train_loader = {InfiniteDataLoader: 385} <ultralytics.data.build.InfiniteDataLoader object at 0x00000169297FFDD0>
-  train_time_start = {float} 1731355033.8569636
-  trainset = {WindowsPath} WindowsPath('C:/Users/derk_/Documents/Universiteit/IRP/Research_Project/Oxford-Flower-dataset/jpg/train')
-  validator = {ClassificationValidator} <ultralytics.models.yolo.classify.val.ClassificationValidator object at 0x000001692AC80BD0>
-  wdir = {WindowsPath} WindowsPath('C:/runs/classify/train53/weights')

# Question
If it is right that `trainer.batch` does not exist (in this case), which other object in trainer - as listed above in the content of trainer - can I extract the current batch?

**Reply by glenn-jocher** [reply]
The `trainer.batch` attribute does not exist in the current implementation. Instead, you can access the current batch through the `train_loader` attribute of the `trainer` object. You can iterate over `trainer.train_loader` to access batches during the `on_train_batch_start` callback.

**Reply by D3RK0** [reply]
@glenn-jocher Many thanks for answering!

Okay, I have been able to access the batches via `trainer.train_loader` and subsequently apply data augmentation to each batch. However, ...
# Problem

... assigning the batch back (see also the code below) to `trainer.train_loader` is not possible, as I get:

`    trainer.train_loader[i] = batch
    ~~~~~~~~~~~~~~~~~~~~^^^
TypeError: 'InfiniteDataLoader' object does not support item assignment`

# Code
```{python}
def batches_augmentation_callback(trainer):
    for i, batch in enumerate(trainer.train_loader):
        images, labels = batch['img'], batch['cls']
        cutmix_or_mixup = LoadOxfordFlowerDataset.data_augmentation(n_classes_in_batch=args.n_classes)
        images, labels = cutmix_or_mixup(images, labels)
        batch = {"img": images, "cls": labels}
        trainer.train_loader[i] = batch     # BUG: trainer.train_loader object does not support item assignment
```

# Question
How can I return the augmented batch to the `trainer` object if assignment does not work?

**Reply by glenn-jocher** [reply]
You cannot directly assign to `trainer.train_loader`. Instead, apply your augmentation within the loop and yield the modified batch back to the training process. Consider using a custom data loader or modifying the dataset directly for more complex augmentations.

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for your interest in Ultralytics 🚀! It seems like you're diving into Ultralytics callbacks for various stages like training, validation, exporting, and prediction. That's a great way to customize your workflows! For comprehensive details, I recommend checking out our [callbacks documentation](https://docs.ultralytics.com/usage/callbacks/) where you'll find everything you need to know about implementing and customizing these callbacks.

If you experience any issues or have specific questions, please feel free to share more details or examples, and be assured that a member of the Ultralytics team will assist you shortly 😊.

In the meantime, upgrade to the latest `ultralytics` package with all requirements to ensure you have the latest features and fixes:

```bash
pip install -U ultralytics
```

Also, consider joining our dynamic community across different platforms for more engaging discussions and support:

- For real-time chat, join us on [Discord](https://ultralytics.com/discord) 🎧.
- If you're interested in in-depth discussions, check out [Discourse](https://community.ultralytics.com).
- For community knowledge sharing, visit our [Subreddit](https://reddit.com/r/ultralytics).

Stay updated with our CI status to ensure smooth operation across different environments:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Is there anything specific you'd like to know or understand better about callbacks? Just let us know, and help is on the way!

**Comment by ruslanraupoff** [comment]
When I'm using single device add_callback works well but multi gpu's not working any solution for this?

**Reply by glenn-jocher** [reply]
@ruslanraupoff for multi-GPU training with callbacks, ensure your callback function is properly registered in all processes. When using DDP mode, callbacks need to be added before the training starts and should be compatible with distributed training environments. If you're experiencing issues, verify that your callback logic is thread-safe and doesn't rely on process-specific resources. For implementation details, refer to our callbacks documentation: https://docs.ultralytics.com/usage/callbacks

**Comment by jnance34** [comment]
Hello, 

I'm trying to implement a custom `on_pretrain_routine_end` callback with a `model.tune()` call, but it appears that the Tuner class isn't adding my custom callback function to the `Tuner.callbacks` list. Here's some pseudo code of what I'm trying to do:

```
def my_custom_callback(trainer):
    print("I'm in my custom callback function!")

model = YOLO("model/yolo11n.pt")
model.add_callback('on_pretrain_routine_end', my_custom_function)
model.tune(...)
# Expect to see "I'm in my custom callback function" for each trial run...
```

For some reason my custom callback function isn't getting called for each training run of the Tuner. Am I missing a step here? I've taken a look at the source code but don't see anything obviously wrong...

Thanks!

-James

**Reply by RizwanMunawar** [reply]
@jnance34 Hi! At the moment, callbacks are not supported for model.tone(...). You can find a complete list of all supported callbacks here: https://docs.ultralytics.com/usage/callbacks/#all-callbacks

**Comment by drbilal216** [comment]
Hi can i and how use callback to freeze initial layers for number of epochs and unfreeze after that, is there any builtin function for that? i used freeze builtin function but it freezes for all the epochs.

**Reply by drbilal216** [reply]
also, after certain epochs the change in model training is minimal can we use callback then to  change lr, how and what parameter to change ?

**Reply by glenn-jocher** [reply]
@drbilal216 yes, you can use callbacks to freeze/unfreeze layers dynamically during training. Here's how to freeze initial layers for specific epochs then unfreeze them:

```python
from ultralytics import YOLO

def dynamic_freeze_callback(trainer):
    freeze_epochs = 10  # Freeze for first 10 epochs
    freeze_layers = 10  # Number of layers to freeze
    
    if trainer.epoch < freeze_epochs:
        # Freeze layers
        for i, (name, param) in enumerate(trainer.model.named_parameters()):
            if i < freeze_layers:
                param.requires_grad = False
    else:
        # Unfreeze all layers
        for param in trainer.model.parameters():
            param.requires_grad = True

model = YOLO("yolo11n.pt")
model.add_callback("on_train_epoch_start", dynamic_freeze_callback)
model.train(data="coco8.yaml", epochs=20)
```

For dynamic learning rate adjustment when training plateaus, you can modify the scheduler within a callback:

```python
def dynamic_lr_callback(trainer):
    if trainer.epoch > 50 and trainer.epoch % 10 == 0:  # Every 10 epochs after epoch 50
        for param_group in trainer.optimizer.param_groups:
            param_group['lr'] *= 0.9  # Reduce LR by 10%

model.add_callback("on_train_epoch_end", dynamic_lr_callback)
```

**Reply by drbilal216** [reply]
@glenn-jocher  Thanks for the code! I'm working on a low-quality PPE dataset. With a custom YAML, I improved YOLOv12n accuracy by 3%, and got similar results on YOLOv11n with better FPS.

The default Nano model lets me use pretrained=True, but custom architectures start from scratch. Still, I can reuse best.pt as pretrained weights on the same YAML later.

I'm now exploring argument tuning and callbacks for further improvements. I also tried freezing/unfreezing layers mid-training but saw no gains.

Lastly, I integrated Flash into YOLOv12, but I don’t see any sign that it’s being used during training (tested on Colab and A100 80GB-server).

**Reply by glenn-jocher** [reply]
@drbilal216 great progress on your PPE dataset improvements! Your approach of using custom YAML configurations and reusing best.pt weights is solid.

Regarding Flash integration with YOLO12, Flash Attention should automatically activate when available, but you can verify it's working by checking GPU memory usage patterns during training - Flash typically shows lower memory consumption. You can also add a simple print statement in your training callback to monitor memory usage and confirm Flash is engaged.
